Cross-cluster painless/execute actions should check permissions only on target remote cluster

quux00 opened the pull request:
Fixes issue #105084, where the cross-cluster painless execute API inappropriately throws a security exception (on a secured cluster) if permissions for the remote index target are not present in the local querying cluster. Since it is a remote request, the security check should be happening on the remote cluster, not the local one.

We were not able to simply implement IndicesRequest.Replaceable to solve this case, as the painless execute API does not support wildcards and implementing Replaceable takes the security checks and index resolutions down paths that don't make sense for this use case.

A new interface, IndicesRequest.SingleIndexNoWildcards was created to support the specific use case that allows remote indices, but no wildcards. Changes were made to RBACEngine and IndiciesAndAliasesResolver to handle this new interface appropriately.

Two new IT tests were added - one covering clusters secured with RCS1 and the other clusters secured with RCS2. Both cover the bug fixed in this commit.

quux00 committed:
Cross-cluster painless/execute actions should check permissions only on target remote cluster

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

elasticsearchmachine commented:
Hi @quux00, I've updated the changelog YAML for you.

quux00 committed:
Example of adding a new Interface for the PainlessExecute use case

quux00 committed:
Added RemoteClusterSecurityPainlessExecuteIT

quux00 commented on x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/RemoteClusterSecurityPainlessExecuteIT.java:
```diff
@@ -0,0 +1,322 @@
+//            Request painlessExecuteRemote = createPainlessExecuteRequest("my_remote_cluster:index1");
+//            Exception exc = expectThrows(Exception.class, () -> performRequestWithRemoteSearchUser(painlessExecuteRemote));
+//            assertThat(exc, notNullValue());
+//            assertThat(exc.getMessage(), containsString("Connection is closed"));  // TODO: is this the expected exception?
```
@jakelandis  - if I uncomment this test case, it receives a "ConnectionClosed" exception rather than a 403 security exception. And worse, the next usage of the client to call to the server (line 219) below, also fails with a ConnectionClosed exception. Any idea why this occurs? 

[Here](https://gist.github.com/quux00/8e5a6f8847f049b23880f44a32643511) is the detailed logs output from the IT test.

You can see there that buried in there is an auth originated error:

```
Caused by: java.lang.AssertionError
	at org.elasticsearch.security@8.13.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationDenialMessages.remoteActionDenied(AuthorizationDenialMessages.java:95)
	at org.elasticsearch.security@8.13.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.lambda$remoteActionDenied$31(AuthorizationService.java:935)
	at org.elasticsearch.security@8.13.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.denialException(AuthorizationService.java:988)
	at org.elasticsearch.security@8.13.0-SNAPSHOT/org.elasticsearch.xpack.security.authz.AuthorizationService.remoteActionDenied(AuthorizationService.java:932)
```

but it seems to cause a fatal error at the Transport layer that seems problematic. I wasn't able to reproduce this problem on my manual testing with a 1.x security set up.

<br/>

And **STDERR output is**:

```
XXX Querying for [local_index]
XXX Querying for [my_remote_cluster:index1]
XXX >>> ABOUT TO UPDATE ROLES

ERROR: Elasticsearch exited unexpectedly, with exit code 1
```

So you can see it fails upon trying to update the roles.

DaveCTurner replied:
The `ConnectionClosed` exception isn't the problem, that's what we'd expect when a node trips an assertion. But tripping an assertion is most definitely a problem. Without having investigated at all, I would guess it's because `cluster:admin/scripts/painless/execute` does not start with `indices:`, but that prefix is required for actions that involve indices and authorization.

quux00 replied:
This same scenario works (returns the expected security_exception) in the RCS1.0 based IT test, but still fails (with the Assertion/ConnectionClosed issue) in the RCS2.0 based IT test, so I plan to follow up with Jake (or someone on the es-security side) to try to get to root cause about what is happening with RCS2.0. It could be test set-up or a bug somewhere in the system.

DaveCTurner replied:
FTR I'm _very_ concerned about this PR having been merged without finding out more about this tripping assertion. From the information I've seen above, I think the most likely explanation is that that this PR itself introduces a bug (or at least violates some assumptions that the security module makes, with currently-unknown consequences).

slobodanadamovic replied:
This indeed is an issue with `cluster:admin/scripts/painless/execute` action being the cluster action (and not indices). It breaks the assumption for RCS 2.0 security, which allows only indices actions to be granted to remote clusters. 

Looking at the code, the painless action seems to be "special" in a sense that it gets [authorized as the indices action](https://github.com/elastic/elasticsearch/pull/85512) when index name is specified.

https://github.com/elastic/elasticsearch/blob/81332093550e6f11b1ecca1ddae345cce06d9c86/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/SecurityActionMapper.java#L46-L51

---

This test is hitting an edge case where request gets [rejected on a local cluster](https://github.com/elastic/elasticsearch/pull/93807) when "user" has no remote index privileges. This is an optimization which avoids sending a request to remote cluster that will anyway be rejected. The assertion is being triggered because this "fail early" approach is executed before `SecurityActionMapper` does the replacement.

I don't see this as a security issue since the request will be rejected with a proper access denied message. But nevertheless, we should handle this case as well.

https://github.com/elastic/elasticsearch/blob/44b0047db5ed516c92a9f5d952319b17f29047c8/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationDenialMessages.java#L95-L99

DaveCTurner replied:
++ thanks for investigating further Slobodan. I'm now less concerned about this specific case, but remain very concerned about the sequence of events that led to this situation. We might not be so lucky next time.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into painless-execute/fix-105084

quux00 committed:
Merge remote-tracking branch 'elastic/main' into painless-execute/fix-105084

quux00 committed:
Merge remote-tracking branch 'elastic/main' into painless-execute/fix-105084

quux00 committed:
spotless fixes to RemoteClusterSecurityPainlessExecuteIT

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

quux00 committed:
Update docs/changelog/105360.yaml

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java:
```diff
@@ -175,6 +176,18 @@ static ResolvedIndices resolveIndicesAndAliasesWithoutWildcards(String action, I
             }
             localIndices.add(IndexNameExpressionResolver.resolveDateMathExpression(name));
         }
+        /// MP TODO: best way to handle this?
+        if (indicesRequest instanceof IndicesRequest.SingleIndexNoWildcards single && single.allowsRemoteIndices()) {
+            assert indices.length == 1;
+
+            // ensure that `index` has a remote name and not a date math expression which includes ':' symbol
+            // since date math expression after evaluation should not contain ':' symbol
+            String indexExpression = IndexNameExpressionResolver.resolveDateMathExpression(indices[0]);
+            if (indexExpression.indexOf(RemoteClusterService.REMOTE_CLUSTER_INDEX_SEPARATOR) >= 0) {
+                return new ResolvedIndices(List.of(), List.of(indices[0]));
+            }
+        }
+
```
I noticed that after the request is successfully authorized on the querying cluster and then forwarded to the fulfilling cluster, there it is again authorized with the same index expression containing again **the index prefixed with the remote cluster**. But that's different from how authz code is expecting things. Authz code expects that the request on the fulfilling cluster to not include the remote cluster name. If they do they are again authz as if they're again a remote index request (but it's a local index request now).

So I think you need to drop the remote cluster name prefix on index names when forwarding to a remote cluster.

albertzaharovits replied:
The code here I think almost works, BUT it uses an uncoventional way to separate remote from local indices, as well as evaluating datemaths for remote index names. I think these might cause subtle problems.

I've put together this diff, which instead reuses the same method to split local/remote indices form expressions that's used for regular indices requests:
```
diff --git a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java
index c8920d0f498..c886f2fde55 100644
--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java
@@ -477,7 +477,7 @@ public class AuthorizationService {
                     resolvedIndicesListener.onResponse(resolvedIndices);
                     return;
                 }
-                final ResolvedIndices resolvedIndices = IndicesAndAliasesResolver.tryResolveWithoutWildcards(action, request);
+                final ResolvedIndices resolvedIndices = indicesAndAliasesResolver.tryResolveWithoutWildcards(action, request);
                 if (resolvedIndices != null) {
                     resolvedIndicesListener.onResponse(resolvedIndices);
                 } else {
@@ -870,8 +870,8 @@ public class AuthorizationService {
         }, listener::onFailure));
     }
 
-    private static String resolveIndexNameDateMath(BulkItemRequest bulkItemRequest) {
-        final ResolvedIndices resolvedIndices = IndicesAndAliasesResolver.resolveIndicesAndAliasesWithoutWildcards(
+    private String resolveIndexNameDateMath(BulkItemRequest bulkItemRequest) {
+        final ResolvedIndices resolvedIndices = indicesAndAliasesResolver.resolveIndicesAndAliasesWithoutWildcards(
             getAction(bulkItemRequest),
             bulkItemRequest.request()
         );
diff --git a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java
index 5fbcc9f5b34..bf1bf7b7d3c 100644
--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java
@@ -28,7 +28,6 @@ import org.elasticsearch.core.Nullable;
 import org.elasticsearch.index.Index;
 import org.elasticsearch.index.IndexNotFoundException;
 import org.elasticsearch.transport.RemoteClusterAware;
-import org.elasticsearch.transport.RemoteClusterService;
 import org.elasticsearch.transport.RemoteConnectionStrategy;
 import org.elasticsearch.transport.TransportRequest;
 import org.elasticsearch.xpack.core.security.authz.AuthorizationEngine;
@@ -47,7 +46,6 @@ import java.util.SortedMap;
 import java.util.concurrent.CopyOnWriteArraySet;
 import java.util.function.Predicate;
 import java.util.function.Supplier;
-import java.util.stream.Stream;
 
 import static org.elasticsearch.xpack.core.security.authz.IndicesAndAliasesResolverField.NO_INDEX_PLACEHOLDER;
 
@@ -121,7 +119,7 @@ class IndicesAndAliasesResolver {
      * @return The {@link ResolvedIndices} or null if wildcard expansion must be performed.
      */
     @Nullable
-    static ResolvedIndices tryResolveWithoutWildcards(String action, TransportRequest transportRequest) {
+    ResolvedIndices tryResolveWithoutWildcards(String action, TransportRequest transportRequest) {
         // We only take care of IndicesRequest
         if (false == transportRequest instanceof IndicesRequest) {
             return null;
@@ -146,7 +144,7 @@ class IndicesAndAliasesResolver {
         return false;
     }
 
-    static ResolvedIndices resolveIndicesAndAliasesWithoutWildcards(String action, IndicesRequest indicesRequest) {
+    ResolvedIndices resolveIndicesAndAliasesWithoutWildcards(String action, IndicesRequest indicesRequest) {
         assert false == requiresWildcardExpansion(indicesRequest) : "request must not require wildcard expansion";
         final String[] indices = indicesRequest.indices();
         if (indices == null || indices.length == 0) {
@@ -163,32 +161,28 @@ class IndicesAndAliasesResolver {
             );
         }
 
+        final ResolvedIndices split;
+        if (indicesRequest instanceof IndicesRequest.SingleIndexNoWildcards single && single.allowsRemoteIndices()) {
+            split = remoteClusterResolver.splitLocalAndRemoteIndexNames(indicesRequest.indices());
+        } else {
+            split = new ResolvedIndices(Arrays.asList(indicesRequest.indices()), List.of());
+        }
+
         // NOTE: shard level requests do support wildcards (as they hold the original indices options) but don't support
         // replacing their indices.
         // That is fine though because they never contain wildcards, as they get replaced as part of the authorization of their
         // corresponding parent request on the coordinating node. Hence wildcards don't need to get replaced nor exploded for
         // shard level requests.
-        final List<String> localIndices = new ArrayList<>(indices.length);
-        for (String name : indices) {
+        final List<String> localIndices = new ArrayList<>(split.getLocal().size());
+        for (String localName : split.getLocal()) {
             // TODO: Shard level requests have wildcard expanded already and do not need go through this check
-            if (Regex.isSimpleMatchPattern(name)) {
-                throwOnUnexpectedWildcards(action, indices);
-            }
-            localIndices.add(IndexNameExpressionResolver.resolveDateMathExpression(name));
-        }
-        /// MP TODO: best way to handle this?
-        if (indicesRequest instanceof IndicesRequest.SingleIndexNoWildcards single && single.allowsRemoteIndices()) {
-            assert indices.length == 1;
-
-            // ensure that `index` has a remote name and not a date math expression which includes ':' symbol
-            // since date math expression after evaluation should not contain ':' symbol
-            String indexExpression = IndexNameExpressionResolver.resolveDateMathExpression(indices[0]);
-            if (indexExpression.indexOf(RemoteClusterService.REMOTE_CLUSTER_INDEX_SEPARATOR) >= 0) {
-                return new ResolvedIndices(List.of(), List.of(indices[0]));
+            if (Regex.isSimpleMatchPattern(localName)) {
+                throwOnUnexpectedWildcards(action, split.getLocal());
             }
+            localIndices.add(IndexNameExpressionResolver.resolveDateMathExpression(localName));
         }
 
-        return new ResolvedIndices(localIndices, List.of());
+        return new ResolvedIndices(localIndices, split.getRemote());
     }
 
     /**
@@ -209,8 +203,8 @@ class IndicesAndAliasesResolver {
         return split;
     }
 
-    private static void throwOnUnexpectedWildcards(String action, String[] indices) {
-        final List<String> wildcards = Stream.of(indices).filter(Regex::isSimpleMatchPattern).toList();
+    private static void throwOnUnexpectedWildcards(String action, List<String> indices) {
+        final List<String> wildcards = indices.stream().filter(Regex::isSimpleMatchPattern).toList();
         assert wildcards.isEmpty() == false : "we already know that there's at least one wildcard in the indices";
         throw new IllegalArgumentException(
             "the action "
```
I think this is a safe localized change to the authz logic to support the `IndicesRequest.SingleIndexNoWildcards` kind of requests.

quux00 replied:
@albertzaharovits 

> I noticed that after the request is successfully authorized on the querying cluster and then forwarded to the fulfilling cluster, there it is again authorized with the same index expression containing again the index prefixed with the remote cluster.

Good catch. I had logic to strip it off on the remote cluster, but you're right, it is better to strip it off before sending it. I've fixed that in the [latest commit](https://github.com/elastic/elasticsearch/pull/105360/commits/93d6ff6f3d59ed4bf57c47a33aa744ca83983a33). 

> I've put together this diff

Thank you very much! I applied that diff. With the above change along with your patch, now the IT test behaves as I expect. Before in one place it was returning a 404 (index not found) where it should return a 403. Now it behaves correctly, I think.

albertzaharovits reviewed (changes requested):
authz on the querying cluster seems OK, but authz on the fulfulling cluster still works as if the request is again a remote one. We need to fix this.

quux00 committed:
First pass at RCS1 based PainlessExecute IT

quux00 committed:
Changed PainlessExecuteAction transport layer to strip off the cluster alias before forwarding it to the remote cluster, as recommended by PR reviewer.

Applied patch from PR review to modify security layer code. Updated security tests to match the new functionality.

Updated the RCS1 and RCS2 IT tests for PainlessExecuteAction to match the new correct behavior.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into painless-execute/fix-105084

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:
```diff
@@ -408,7 +408,12 @@ public void authorizeIndexAction(
 
     private static boolean allowsRemoteIndices(TransportRequest transportRequest) {
-        return transportRequest instanceof IndicesRequest.Replaceable replaceable && replaceable.allowsRemoteIndices();
+        /// MP TODO: best way to handle this?
```
nit: you may reference this new issue in the comment here: https://github.com/elastic/elasticsearch/issues/105598

quux00 replied:
Done. Thanks for the link.

albertzaharovits reviewed (approved):
LGTM

DaveCTurner commented on x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/RemoteClusterSecurityRCS2PainlessExecuteIT.java:
```diff
@@ -0,0 +1,321 @@
+            assertThat(responseBody, equalTo("{\"result\":[\"test\"]}"));
+        }
+        {
+            // // TEST CASE 2: Query remote cluster for index1 - should fail since no permissions granted for remote clusters yet
```
Let's decide what to do with these test cases before merging, rather than just leaving them commented out like this. If we care about them, let's get them working (or at least open an issue to report the problem they represent). If we don't care about them, let's just remove them.

DaveCTurner commented on x-pack/plugin/security/qa/multi-cluster/src/javaRestTest/java/org/elasticsearch/xpack/remotecluster/RemoteClusterSecurityRCS1PainlessExecuteIT.java:
```diff
@@ -0,0 +1,276 @@
+        }
+        {
+            // // TEST CASE 2: Query remote cluster for index1 - should fail since no permissions granted for remote clusters yet
+            // Request painlessExecuteRemote = createPainlessExecuteRequest("my_remote_cluster:index1");
+            // Exception exc = expectThrows(Exception.class, () -> performRequestWithRemoteSearchUser(painlessExecuteRemote));
+            // assertThat(exc, notNullValue());
+            // assertThat(exc.getMessage(), containsString("Connection is closed")); // TODO: is this the expected exception?
+
+            // assertThat(exc.getResponse().getStatusLine().getStatusCode(), is(403));
+            // String errorResponseBody = EntityUtils.toString(exc.getResponse().getEntity());
+            // assertThat(err orResponseBody, containsString("unauthorized for user [remote_search_user]"));
+            // assertThat(errorResponseBody, containsString("on indices [secretindex]"));
+            // assertThat(errorResponseBody, containsString("\"type\":\"security_exception\""));
+            //
+            // Response response = performRequestWithRemoteSearchUser(painlessExecuteRemote);
+            // String responseBody = EntityUtils.toString(response.getEntity());
+            // assertOK(response);
+            // assertThat(responseBody, equalTo("{\"result\":[\"test\"]}"));
```
Let's decide what to do with these test cases before merging, rather than just leaving them commented out like this. If we care about them, let's get them working (or at least open an issue to report the problem they represent). If we don't care about them, let's just remove them.

quux00 replied:
Yes, I'm still working on the tests with @slobodanadamovic

quux00 replied:
Tests are updated now.

DaveCTurner reviewed (commented):
I think the commented-out test cases need addressing before we can merge this, particularly if we're commenting them out because the changes here lead to a tripping assertion.

DaveCTurner reviewed (dismissed):
Sorry [previous review](https://github.com/elastic/elasticsearch/pull/105360#pullrequestreview-1887246485) was supposed to be a "request changes" one but I forgot to press the button.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into painless-execute/fix-105084

quux00 committed:
Finalized changes to the RCS1 and RCS2 based IT tests

quux00 committed:
Updated placeholder TODO comments in code with issue links and proper javadoc for SingleIndexNoWildcards

elasticsearchmachine commented:
Hi @quux00, I've updated the changelog YAML for you.

quux00 committed:
Update docs/changelog/105360.yaml

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.12  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 105360`

quux00 commented:
Decided against the backport since it had several conflicts and were are past time to make the 8.12.2 cutoff.
